src/copy.c

```diff
@@ -184,7 +184,7 @@ TSCopyMultiInsertBufferInit(ChunkInsertState *cis, Point *point)
 	TSCopyMultiInsertBuffer *buffer;
 
 	buffer = (TSCopyMultiInsertBuffer *) palloc(sizeof(TSCopyMultiInsertBuffer));
-	memset(buffer->slots, 0, sizeof(TupleTableSlot *) * MAX_BUFFERED_TUPLES);
+	memset((void *) buffer->slots, 0, sizeof(TupleTableSlot *) * MAX_BUFFERED_TUPLES);
```

Comment:
Why does it warn without the cast here? Implicit cast from (or to) an arbitrary pointer type (unless different cv qualifier) to `void*` is well-defined in the standard.

Reply:
Most of the clang-tidy warnings are about pretty well-defined things that are considered error-prone. This warning is specifically about implicit casts of multi-level pointers, they consider it easy to mismatch the indirection levels there. Happened to me as well, idk how useful it is.

https://releases.llvm.org/18.1.0/tools/clang/tools/extra/docs/clang-tidy/checks/bugprone/multi-level-implicit-pointer-conversion.html

Reply:
I think these warnings are misguided, since uses of `void*` are used precisely when dealing with pointers to raw memory area and all bets are off.

This comment, for example:
> This results in an implicit pointer level conversion, which could cause issues if foo() dereferences the pointer assuming it’s a int* pointer.

If the author of `foo(void* ptr)` expects an `int*` pointer, they should declare it as `foo(int*)`, not blindly assume that it will be called correctly, so this is an error in the declaration of `foo`, not in the way `foo` is used.

It is easier and far more common to introduce bugs because of over-using casts (in general, but specifically when casting to `void*`). For example, these will be bugs that are hard to find if the cast is there, but caught by the compiler if they are not:

This has mixed up the order of the parameters, but this is hidden by the cast:
```
void foo(char *ap, const char *bp, size_t len) {
  ...
  memcpy((void*) bp, (void*) ap, len);
  ...
}
```

This has missed including `stdlib.h` but is hidden because of the cast. This could cause a crash if `malloc` is defined as a macro. (There are some [discussions on StackOverflow](https://stackoverflow.com/questions/59707204/should-do-i-cast-a-void-pointer-in-c-in-general) on this.)
```c
void main() {
  char* buffer = malloc(100);
  ...
```

This returns a buffer of `double` but accidentally treats it as `int`. This could be a result of a change of `get_data` and will be hidden because of the cast, possibly not even triggering a bug in the regression suite:

```c
double *get_data(const char *filename);

void my_reader(const char *filename) {
  int *buffer = (int *)malloc(100);
  ...
  memcpy(buffer, (void *)get_data(filename), 100);
  ...
}
```

I realize that it is possible to make mistakes when using implicit casts to `void*`, but I think they are far less common than variations of the above.